Automattic/wp-calypso
# Checkout: Change existingCardProcessor to get postalCode/countryCode from contactDetails

#### Changes proposed in this Pull Request

Because the upsell modal does not offer contact details fields, and because it needs to provide postal code and country directly from the metadata stored with the card, we therefore need to wait until the card is selected before we know the values of that data. As a result, the `existingCardProcessor` is unique among card processors in that it requires postal code and country to be submitted as part of the transaction data rather than extracting it from the contact details.

Aside from the increased complexity of this special case, this means that the `existingCardProcessor` has to validate the existence of these fields, providing a custom (untranslated) error message to the user whereas the other payment processors can rely on the server-side logic of the transactions endpoint to handle this data. This is more problematic since we're beginning to support countries which do not have postal codes, meaning we'll need to add detection and support for these countries inside the payment processor.

This PR attempts to resolve this complexity by making the `existingCardProcessor` work like other payment processors, getting its postal code and country from the contact details and removing the client-side validation in favor of the server-side logic.

The biggest challenge of doing this is what was described above: if we are not using a contact form, we do not actually know these contact details until the card is selected in checkout, and we need that data as we're preparing the `CheckoutProvider` which has no easy way to determine the selected card. Fortunately, the upsell also does not offer a card selector; it always uses the first available card, so we do in fact know which card will be selected before we create the form.

My only hesitation about doing this is that it does make the `existingCardProcessor` less flexible; for example if the upsell (or some other implementation) wanted to get the postal code and country from the saved card and did offer a card selector, then we'd be stuck, but I think that's premature optimization at this point and the decreased complexity seems worth the possible risk. Let me know if you disagree.

#### Testing instructions

##### To test checkout:

- You'll need to have a saved card. If you do not have one already you can add one by visiting`/me/purchases/add-payment-method`.
- Add a product to your cart and visit checkout.
- Complete the form and select a saved card as the payment method.
- Submit the form.
- Verify that the payment is successful without any errors.


##### To test the one-click upsell:

- To get the upsell to appear you'll need a Personal or Premium plan already purchased, and you'll need to have at least one stored card, then visit `/checkout/offer-quickstart-session/1234/example.com` where `example.com` is your site (the number is a receipt ID but doesn't really matter so you can just make it up).
- Click "Yes, I want a WordPress Expert by my side!".
- Verify that the one-click modal appears (if you're instead redirected to checkout, then you probably have an invalid postal code saved with that card; you can fix this by adding a new card at `/me/purchases/add-payment-method`).
- Verify that the payment is successful without any errors.
